### ISSUE TITLE
Another Python OpenMfx host implementation & feedback

### ISSUE THREAD

**Issue by follower** [eliemichel/PythonMfx#3]
~Guess that's what happens when I take six months to make a project public! :/~

~Only just~ Today I re-started on "tidying up" a project I did back in June/July creating a `ctypes`-based Python wrapper for OpenMfx:

 * https://gitlab.com/RancidBacon/openmfx-python-ctypes-poc

When completed it will be a second, independently developed, Python OpenMfx host implementation, which IIRC is sufficient for standards consideration. :D

~Still, guess it's better to discover `PythonMfx` now than _after_ I'd finished tidying it up...~

### Anyway...

In brief, when I discovered OpenMfx (about 6 months ago, I guess...) I thought it was a super cool concept.

I stumbled upon OpenMfx after playing with Blender's Geometry Nodes & wanting to write a Geometry Node in Python--only to discover that's not possible.

I saw that you'd created a special build of Blender with OpenMfx support but thought that anything that requires a special download would _really_ hurt adoption of OpenMfx. So...

For some reason (I think it's because I hate compilation :D ), I'm a _huge_ fan of the power of FFI/libffi/`ctypes` in higher level/dynamic/scripting languages (e.g. Python, gdscript) and knew it should be possible to make OpenMfx work (at _some_ level) within Blender purely with Python+`ctypes`. So thought I'd work on a proof-of-concept to demonstrate that approach for you.

And, I _did_ get something working.

But, well, too many projects & too high standards for "publishing" and six months go by...

However, that's history now, so, looking forward & in the hope of extracting something useful...

### Suggestions to Assist OpenMfx Host Developers

Based on my experience implementing an OpenMfx host, I would strongly encourage you to consider these suggestions in order to increase OpenMfx adoption:

 1. **Create/modify an OpenMfx plugin to have defaults that result in a non-zero size mesh** in order to reduce the amount of API surface area that needs to be implemented before seeing visual results.
 2. **Create/modify a OpenMfx plugin that can cope with null pointers in suite structs** & logs a warning rather than crashing (in order to reduce required initial API surface area implementation requirements).
 3. **Add JSON input & output to a Python+`ctypes`-based host** so that, as a first step, someone doesn't have to implement _any_ of the API in order to start integrating an OpenMfx plugin into their application--they can just use `stdin`/`stdout` , subprocesses and/or files.

(As an aside, while it's a definite PITA to implement OpenFx-based plugin architecture due to, well, C & the inherent constraints of FFI & the age of the standard, I _do_ think it was a good decision to base OpenMfx on the existing architecture for a variety of reasons that I think probably outweigh the downsides.)

### Target Projects

FWIW these were the languages/applications I had/have plans to create at least proof-of-concept level support for and/or encourage to consider incorporating OpenMfx:

 * Languages:
    * Python: https://gitlab.com/RancidBacon/openmfx-python-ctypes-poc
    * gdscript
    * Rust: https://gitlab.com/RancidBacon/openmfx-rust-poc
 * Applications
   * Blender
   * Blackjack
   * Fornjot
   * Inkscape

[_I'll endeavour to update this comment and/or issue with some more details but, ah, well, ya know..._]

**TL;DR: OpenMfx is super cool, I encourage you to continue with its development.**

**Comment by eliemichel** [comment]
Thanks a lot for this detailed message! I understand the frustration of discovering a duplicate of your work ^^ But indeed it's not lost since having independent implementations of the same thing is always very instructive :)

Glad you are working on a Rust PoC, this is something I had in mind as well though I did not start anything so go ahead!

You mentioned having something that works regarding running OpenMfx effects in a standard Blender? Do you have more feedback about this?

Anyways very glad to see you appreciate the project! :)
